### DISCUSSION TITLE
Localizable Strings in Custom Controls

### DISCUSSION THREAD

**Discussion by rcopstein** [Flinesoft/BartyCrouch#228] [Q&A] [answered]
In my project I have a custom control that subclasses NSView and contains some customizable text fields. These text fields are bound to text properties that are @IBInspectable.

The same way that BartyCrouch generates localization strings for standard controls (such a a text field getting a "<storyboard-id>.title" key in the .strings file), how can I get the same behaviour on my own controls? It's not a static value that changes with language, I want an exposed parameter to "trigger" the generation of a key in the .strings file.

Does anyone know how to do this?

**Comment by Jeehut** [comment, marked as answer]
@rcopstein Are you talking about triggering the generation of a key in the `.strings` file of a XIB file? Or are you talking about the `Localizable.strings` file? If it's the latter, that's the `transform` option, see the README. If you're asking for the first, the only way currently to add a key to a `.strings` file is by adding it manually to the source languages you have specified in the BartyCrouch configuration (defaults to `en` if you didn't set one). Then, the key will be automatically added to all other languages.

There is no feature that will transform a key in code to a key in a XIB file, if that's what you're asking for. But I'm not sure if I understood you correctly. Feel free to provide more context, some code or even a small demo project.

**Reply by rcopstein** [reply]
@Jeehut Thank you for the answer. I should have given more details, my bad. I'm working on a macOS app using storyboards that are marked as localizable. Every time I add a component such as NSButton or NSTextField, BartyCrouch adds a key to the `Storyboard.strings` for that storyboard in all the supported languages. It's basically "ID of the control.Property" that gets added and bound to most NS controls.

What I am doing, instead of adding an NSButton, is adding a custom control, that is, something that inherits NSView. I was wondering if there is a way to make my custom control trigger the generation of a key when it is added to a storyboard, much like the NS controls do?

**Reply by Jeehut** [reply]
@rcopstein BartyCrouch actually doesn't parse any Interface Builder files, instead it uses the `extractLocStrings` command shipped as part of Xcode. So whatever that tool supports works currently for generating keys and it seems it doesn't support your use case.

Having that said, I am considering writing my own parser for the next major version of BartyCrouch where we can add features like this, but I can't promise anything there. I might be able to get it done in the next 12 months, but it's a big change and I'm not even sure if I will be happy with the performance or quality at all, we'll see.

You could move the translation part to code in the meantime and use the supported features there.

**Reply by rcopstein** [reply]
Got it! Yeah, I've been making do with `NSLocalizedString` in the meantime. It's not a bad workflow (especially when compared to the one without BartyCrouch), but just wondering if it did get better by providing what I mentioned.

Looking forward to seeing where you are going next! 😉
